File: Stocker/ProjetoLDS/Stocker/Backend/StockerAPI/StockerAPI/Models/Dto/typeStatisticsDto.cs
namespace StockerAPI.Models.Dto
{
    /// <summary>
    /// Representa as estatísticas de um tipo de produto, incluindo o total gasto, quantidade comprada, número de compras e a média gasta.
    /// </summary>
    public class TypeStatisticsDto
    {
        /// <summary>
        /// Tipo de produto.
        /// Indica a categoria ou tipo do produto, como "Frutas", "Laticínios", etc.
        /// </summary>
        public string ProductType { get; set; }

        /// <summary>
        /// Total gasto em compras desse tipo de produto.
        /// Representa o valor total gasto em todas as compras do tipo de produto.
        /// </summary>
        public double TotalSpent { get; set; }

        /// <summary>
        /// Quantidade total comprada do tipo de produto.
        /// Representa a quantidade total de produtos adquiridos para o tipo de produto.
        /// </summary>
        public double TotalQuantity { get; set; }

        /// <summary>
        /// Número total de compras realizadas para o tipo de produto.
        /// Indica quantas vezes o tipo de produto foi comprado.
        /// </summary>
        public int TotalPurchases { get; set; }

        /// <summary>
        /// Média gasta por compra no tipo de produto.
        /// Representa o valor médio gasto em cada compra do tipo de produto.
        /// </summary>
        public double AverageSpent { get; set; }
    }
}

File: Stocker/ProjetoLDS/Stocker/Frontend/StockerMobileUI/stocker_web_ui/lib/models/typeStatisticsDto.dart
/// Representa as estatísticas relacionadas a um tipo de produto, incluindo dados financeiros e de consumo.
///
/// Esta classe é usada para modelar informações como o tipo de produto, total gasto, quantidade adquirida,
/// número de compras realizadas e a média de gastos associados ao tipo de produto.
class TypeStatisticsDto {
  /// Tipo do produto.
  final String productType;

  /// Valor total gasto no tipo de produto.
  final double totalSpent;

  /// Quantidade total adquirida do tipo de produto.
  final double totalQuantity;

  /// Número total de compras realizadas para o tipo de produto.
  final int totalPurchases;

  /// Gasto médio por compra relacionado ao tipo de produto.
  final double averageSpent;

  /// Construtor da classe [TypeStatisticsDto].
  ///
  /// - [productType]: Tipo do produto (obrigatório).
  /// - [totalSpent]: Valor total gasto no tipo de produto (obrigatório).
  /// - [totalQuantity]: Quantidade total adquirida do tipo de produto (obrigatório).
  /// - [totalPurchases]: Número total de compras realizadas (obrigatório).
  /// - [averageSpent]: Valor médio gasto por compra (obrigatório).
  TypeStatisticsDto({
    required this.productType,
    required this.totalSpent,
    required this.totalQuantity,
    required this.totalPurchases,
    required this.averageSpent,
  });

  /// Converte um mapa JSON numa instância de [TypeStatisticsDto].
  factory TypeStatisticsDto.fromJson(Map<String, dynamic> json) {
    return TypeStatisticsDto(
      productType: json['productType'] as String,
      totalSpent: (json['totalSpent'] as num).toDouble(),
      totalQuantity: (json['totalQuantity'] as num).toDouble(),
      totalPurchases: json['totalPurchases'] as int,
      averageSpent: (json['averageSpent'] as num).toDouble(),
    );
  }

  /// Converte uma instância de [TypeStatisticsDto] para um mapa JSON.
  Map<String, dynamic> toJson() {
    return {
      'productType': productType,
      'totalSpent': totalSpent,
      'totalQuantity': totalQuantity,
      'totalPurchases': totalPurchases,
      'averageSpent': averageSpent,
    };
  }
}
